libs/core/synchronization/include/hpx/synchronization/detail/range_lock_impl.hpp

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include <hpx/execution_base/this_thread.hpp>
+
+#include <atomic>
+#include <boost/container/flat_map.hpp>
+#include <cstddef>
+#include <utility>
+#include <vector>
+
+namespace hpx::synchronization::detail {
+
+    template <typename Lock, template <typename> typename Guard>
+    class RangeLock
+    {
+        template <typename Key, typename Value>
+        using MapTy = boost::container::flat_map<Key, Value>;
+
+        Lock mtx;
+        std::size_t counter = 0;
+        MapTy<std::size_t, std::pair<std::size_t, std::size_t>> rangeMap;
+        MapTy<std::size_t, std::shared_ptr<std::atomic_bool>> waiting;
+
+    public:
+        std::size_t lock(std::size_t begin, std::size_t end);
+        std::size_t try_lock(std::size_t begin, std::size_t end);
+        void unlock(std::size_t lockId);
+    };
```

Comment:
> Usually, mutex types are movable, but not copyable.
Sure, just need to delete the copy constructor and define a move constructor for it I believe. Do I need to worry about thread safety while moving? like say another thread trying to lock? 

> This will most likely also allow to replace the std::shared_ptr<std::atomic_bool> with a simpler solution based on unique_ptr.
I am unsure how adding move constructor would help change the flag from `shared_ptr` to `unique_ptr`. 
To summarise, if a thread locks range [0, 20), it also creates an `atomic_bool` flag. 
So if another thread (2) wants to lock an intersecting range, say [10, 30). It spins on the `atomic_bool` flag before trying to acquire the lock again, trying to acquire the range_lock is an expensive operation which we want to avoid unless we are confident the range_lock can be acquired.

When performing unlock, we delete the atomic_bool flag while the other threads might still be spinning on it, using `unique_ptr` would delete the atomic_bool causing a segmentation fault, `shared_ptr` will not cause this issue.